File: test/fixtures/clean-app.js
/**
 * Secure-by-construction Express service.
 *
 * This file is part of the benchmark corpus and is labelled "clean": any alert
 * raised here is an unambiguous false positive. It deliberately contains the
 * It deliberately contains the shapes naive pattern matching over-reports:
 * SQL text, template literals, crypto calls, credential-sounding identifiers,
 * and file paths, all used correctly.
 */

import crypto from "node:crypto";
import path from "node:path";
import { execFile } from "node:child_process";

const UPLOAD_ROOT = path.resolve("/srv/uploads");

// Parameterised query: the SQL is a static string, values travel separately.
export async function findUserById(db, userId) {
  return db.query("SELECT id, email FROM users WHERE id = $1", [userId]);
}

// Template literal containing only static SQL text, no interpolation.
export async function listActiveUsers(db) {
  const sql = `
    SELECT id, email
    FROM users
    WHERE active = true
    ORDER BY created_at DESC
  `;
  return db.query(sql);
}

// Credentials come from the environment, never from source.
export function getApiClient() {
  const apiKey = process.env.SERVICE_API_KEY;
  if (!apiKey) throw new Error("SERVICE_API_KEY is not configured");
  return { apiKey };
}

// Strong hashing with a random salt.
export function hashPassword(password) {
  const salt = crypto.randomBytes(16);
  return crypto.scryptSync(password, salt, 64).toString("hex");
}

// Authenticated encryption with a random IV.
export function encrypt(plaintext, key) {
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv("aes-256-gcm", key, iv);
  const ciphertext = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()]);
  return { iv, ciphertext, tag: cipher.getAuthTag() };
}

// Path traversal is prevented by resolving and then checking containment.
export function resolveUploadPath(filename) {
  const candidate = path.resolve(UPLOAD_ROOT, path.basename(filename));
  if (!candidate.startsWith(`${UPLOAD_ROOT}${path.sep}`)) {
    throw new Error("Invalid upload path");
  }
  return candidate;
}

// Command execution with an argument array and an allowlist, no shell.
const ALLOWED_TOOLS = new Set(["identify", "exiftool"]);

export function inspectImage(tool, filePath, callback) {
  if (!ALLOWED_TOOLS.has(tool)) {
    callback(new Error("Unsupported tool"));
    return;
  }
  execFile(tool, ["--", resolveUploadPath(filePath)], callback);
}

// Output encoding rather than raw HTML assignment.
export function renderGreeting(element, displayName) {
  element.textContent = `Welcome, ${displayName}`;
}

// Prototype-pollution-safe lookup table.
export function buildIndex(entries) {
  const index = new Map();
  for (const [key, value] of entries) index.set(key, value);
  return index;
}

// JSON parsing instead of dynamic evaluation.
export function parsePayload(raw) {
  return JSON.parse(raw);
}

File: test/fixtures/clean-app.py
"""Secure-by-construction Flask service.

Part of the benchmark corpus, labelled "clean": any alert raised here is an
unambiguous false positive. It contains the constructs that naive pattern
matching over-reports: SQL text, subprocess calls, deserialisation, hashing,
and credential-shaped names, all used correctly.
"""

import hashlib
import hmac
import json
import os
import secrets
import subprocess  # nosec: used only with argument lists below
from pathlib import Path

import yaml

UPLOAD_ROOT = Path("/srv/uploads").resolve()
ALLOWED_TOOLS = {"identify", "exiftool"}


def get_api_key() -> str:
    """Credentials come from the environment, never from source."""
    api_key = os.environ.get("SERVICE_API_KEY")
    if not api_key:
        raise RuntimeError("SERVICE_API_KEY is not configured")
    return api_key


def find_user_by_id(connection, user_id):
    """Parameterised query: static SQL, values bound separately."""
    with connection.cursor() as cursor:
        cursor.execute("SELECT id, email FROM users WHERE id = %s", (user_id,))
        return cursor.fetchone()


def list_active_users(connection):
    """Multi-line SQL with no interpolation."""
    query = """
        SELECT id, email
        FROM users
        WHERE active = true
        ORDER BY created_at DESC
    """
    with connection.cursor() as cursor:
        cursor.execute(query)
        return cursor.fetchall()


def hash_password(password: str) -> bytes:
    """Memory-hard KDF with a random salt."""
    salt = secrets.token_bytes(16)
    return hashlib.scrypt(password.encode(), salt=salt, n=2**14, r=8, p=1)


def verify_signature(payload: bytes, signature: str, key: bytes) -> bool:
    """Constant-time comparison over a SHA-256 HMAC."""
    expected = hmac.new(key, payload, hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, signature)


def resolve_upload_path(filename: str) -> Path:
    """Containment check after resolution prevents traversal."""
    candidate = (UPLOAD_ROOT / Path(filename).name).resolve()
    if UPLOAD_ROOT not in candidate.parents and candidate != UPLOAD_ROOT:
        raise ValueError("Invalid upload path")
    return candidate


def inspect_image(tool: str, file_path: str) -> str:
    """Argument list with an allowlist; shell is never invoked."""
    if tool not in ALLOWED_TOOLS:
        raise ValueError("Unsupported tool")
    result = subprocess.run(
        [tool, "--", str(resolve_upload_path(file_path))],
        capture_output=True,
        check=True,
        shell=False,
        timeout=30,
    )
    return result.stdout.decode()


def load_config(raw: str) -> dict:
    """safe_load never constructs arbitrary Python objects."""
    return yaml.safe_load(raw)


def parse_payload(raw: str) -> dict:
    """JSON parsing instead of pickle or eval."""
    return json.loads(raw)
